File: code/assignment/week03/no4_java/Headquarter.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Headquarter {
    private String type;
    private int totalStrength;
    private String[] warriorNameArray;
    private List<Warrior> warriorList;
    // 各种武士对应的数量
    private Map<String, Integer> countWarriors;
    // 是否还能制造武士
    private boolean active;

    public Headquarter(String type, int totalStrength, String[] warriorNameArray) {
        this.type = type;
        this.totalStrength = totalStrength;
        this.warriorNameArray = warriorNameArray;

        warriorList = new ArrayList<>();

        this.countWarriors = new HashMap<>();
        for (String name : warriorNameArray) {
            this.countWarriors.put(name, 0);
        }

        this.active = true;
    }

    Warrior createWarrior(String type, int strength) {
        if (this.totalStrength > strength) {
            Warrior warrior = new Warrior(type, strength);

            this.warriorList.add(warrior);
            this.totalStrength -= strength;
            this.countWarriors.put(type, this.countWarriors.get(type) + 1);
            return warrior;
        } else {
            return null;
        }
    }

    public String getType() {
        return type;
    }

    public int getTotalStrength() {
        return totalStrength;
    }

    public Map<String, Integer> getCountWarriors() {
        return countWarriors;
    }

    public boolean isActive() {
        return active;
    }

    public void setActive(boolean active) {
        this.active = active;
    }
}

File: code/assignment/week03/no4/Headquarter.cpp
#include "Warrior.cpp"
#include <map>
#include <string>
#include <vector>

using namespace std;

// 司令部
class Headquarter {
	string type;
	int totalStrength;
	string warriorNameArray[5] = { 0 };
	vector<Warrior> warriorList;
	// 各种武士对应的数量
	map<string, int> countWarriors;
	// 是否还能制造武士
	bool active;

public:
	Headquarter(string type, int totalStrength, string warriorNameArray[5]);
	bool createWarrior(const string &type, int strength, Warrior &warrior);
	string getType();
	int getTotalStrength();
	map<string, int> getCountWarriors();
	bool isActive();
	void setActive(bool active);
};

Headquarter::Headquarter(string type, int totalStrength,
		string warriorNameArray[]) {
	this->type = type;
	this->totalStrength = totalStrength;

	for (int i = 0; i < 5; i++) {
		this->warriorNameArray[i] = warriorNameArray[i];
		this->countWarriors.insert( { warriorNameArray[i], 0 });
	}
	this->active = true;
}

bool Headquarter::createWarrior(const string &type, int strength,
		Warrior &warrior) {
	if (this->totalStrength > strength) {
		Warrior warrior(type, strength);
		this->warriorList.push_back(warrior);
		this->totalStrength -= strength;
		this->countWarriors.insert(
				pair<string, int>(type,
						this->countWarriors.find(type)->second + 1));
		return true;
	} else {
		return false;
	}
}

string Headquarter::getType() {
	return this->type;
}

int Headquarter::getTotalStrength() {
	return this->totalStrength;
}

map<string, int> Headquarter::getCountWarriors() {
	return this->countWarriors;
}

bool Headquarter::isActive() {
	return this->active;
}

void Headquarter::setActive(bool active) {
	this->active = active;
}
